transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/jobscheduling/Uploader.java

```diff
@@ -14,13 +14,107 @@
 
+  private final int LOCK_TIMEOUT = 10000;
+  private final Context context;
+  private final BackendRegistry backendRegistry;
+  private final EventStore eventStore;
+  private final WorkScheduler workScheduler;
+  private final Executor executor;
+  private final SynchronizationGuard guard;
+
   @Inject
-  public Uploader() {}
+  public Uploader(
+      Context context,
+      BackendRegistry backendRegistry,
+      EventStore eventStore,
+      WorkScheduler workScheduler,
+      Executor executor,
+      SynchronizationGuard guard) {
+    this.context = context;
+    this.backendRegistry = backendRegistry;
+    this.eventStore = eventStore;
+    this.workScheduler = workScheduler;
+    this.executor = executor;
+    this.guard = guard;
+  }
+
+  boolean isNetworkAvailable() {
+    ConnectivityManager connectivityManager =
+        (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+    NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
+    return activeNetworkInfo != null && activeNetworkInfo.isConnected();
+  }
+
+  void upload(String backendName, int attemptNumber, Runnable callback) {
+    this.executor.execute(
+        () -> {
+          try {
+            TransportContext transportContext =
+                TransportContext.builder().setBackendName(backendName).build();
+            if (!isNetworkAvailable()) {
+              guard.runCriticalSection(
+                  LOCK_TIMEOUT,
+                  () -> {
+                    workScheduler.schedule(transportContext, attemptNumber + 1);
+                    return null;
+                  });
+            } else {
+              logAndUpdateState(transportContext, attemptNumber);
+            }
+          } finally {
+            callback.run();
+          }
+        });
+  }
+
+  void logAndUpdateState(TransportContext transportContext, int attemptNumber) {
+    TransportBackend backend = backendRegistry.get(transportContext.getBackendName());
+    List<EventInternal> eventInternals = new ArrayList<>();
+    Iterable<PersistedEvent> persistedEvents =
+        guard.runCriticalSection(LOCK_TIMEOUT, () -> eventStore.loadAll(transportContext));
 
-  void upload(String backendName, int attemptNumber, Runnable callback) {}
+    // Donot make a call to the backend if the list is empty.
+    if (!persistedEvents.iterator().hasNext()) {
+      return;
+    }
+    // Load all the backends to an iterable of event internals.
+    for (PersistedEvent persistedEvent : persistedEvents) {
+      eventInternals.add(persistedEvent.getEvent());
+    }
+    BackendResponse response = backend.send(BackendRequest.create(eventInternals));
+    guard.runCriticalSection(
+        LOCK_TIMEOUT,
+        () -> {
+          if (response.getStatus() == BackendResponse.Status.OK) {
+            eventStore.recordSuccess(persistedEvents);
+            eventStore.recordNextCallTime(transportContext, response.getNextRequestWaitMillis());
```

Comment:
this looks incorrect, instead of persisting the offset e.g. `2 minutes` you need to persist `clock.now() + 2 minutes`